quint/mempool/Consensus.qnt

```diff
@@ -0,0 +1,203 @@
+// -*- mode: Bluespec; -*-
+// The common consensus abstraction for blockchain where processes agree on a block (that is, a set of transactions).
+
+module Consensus{
+
+    import Spells.* from "Spells"
+    import System.* from "System"
+
+    type ConsensusState = {
+      proposed: Process -> Set[Tx],
+      decided: Process -> Set[Tx]
+    }
+
+    pure def newConsensus(P: Set[Process]): ConsensusState = {
+      proposed: P.mapBy(p => Set()),
+      decided: P.mapBy(p => Set())
+    }
+
+    // 1. helpers
+
+    pure def processes(state: ConsensusState): Set[Process] = {
+      state.proposed.keys()
+    }
+
+    pure def hasProposed(state: ConsensusState, p: Process): bool = {
+      state.proposed.get(p).size() > 0
+    }
+    
+    pure def hasDecided(state: ConsensusState, p: Process): bool = {
+      state.decided.get(p).size() > 0
+    }
+
+    pure def isProposed(state: ConsensusState, t: Tx): bool = {
+      state.proposed.keys().filter(p => state.proposed.get(p).contains(t)).size() > 0
+    }
+
+    pure def isDecided(state: ConsensusState, t: Tx): bool = {
+      state.decided.keys().filter(p => state.decided.get(p).contains(t)).size() > 0
+    }
+
+    pure def hasDecision(state: ConsensusState): bool = {
+      state.decided.keys().filter(p => state.decided.get(p).size()>0).size() > 0
+    }
+
+    pure def decisionOf(state: ConsensusState, p: Process): Set[Tx] = {
+      state.decided.get(p)
+    }
+
+    pure def proposalOf(state: ConsensusState, p: Process): Set[Tx] = {
+      state.proposed.get(p)
+    }
+
+    pure def getProposed(state: ConsensusState): Set[Tx] = {
+      flatten(mapValues(state.proposed))
+    }
+
+    pure def getDecided(state: ConsensusState): Set[Tx] = {
+      flatten(mapValues(state.decided))
+    }
+
+    pure def isSupersededBy(state0: ConsensusState, state1: ConsensusState): bool = {
+      and {
+        state0.processes() == state1.processes(),
+	state0.processes().forall(
+	  p =>
+	    or {
+	      not(state0.hasDecided(p)) and state1.hasDecided(p),
+	      state0.hasDecided(p) and state1.hasDecided(p)
+	    }
+        )
+      }
+    }
+
+    // 2. preconditions
+
+    pure def mayPropose(state: ConsensusState, p: Process, txs: Set[Tx]): bool = {
+      state.processes().contains(p) and not(state.hasProposed(p)) and txs.size() != 0
```

Comment:
The `txs.size() != 0` constraint does not exist in production (validators may propose empty blocks)

Reply:
Indeed. This is to avoid stuttering situations, in which only empty sets are decided. As this specification is a high level one, I have made such a simplification.  (To have progress, we would need that infinitely often non-empty sets are decided.)